### ISSUE TITLE
Make compiler minimally re-entrant

### ISSUE THREAD

**Issue by jfecher** [noir-lang/noir#4592]
### Problem

Once https://github.com/noir-lang/noir/issues/4589 is complete, the compiler will be re-entrant, but not necessarily minimally so. For example, an easy way to make the compiler re-entrant would be to restart name resolution and type checking for the entire current code-base, but this is obviously not always necessary.

### Happy Case

The compiler should be minimally re-entrant, so that it only repeats work when it needs to. Local insertions of code should not require global updates.

### Project Impact

None

### Workaround

None

### Would you like to submit a PR for this Issue?

None

**Comment by jfecher** [comment]
We've moved to using an elaborator over having the compiler loop internally, so this is no longer needed.

**Comment by jfecher** [comment]
We've moved to using an elaborator over having the compiler loop internally, so this is no longer needed.
